README.md
<p align="center">
  <a href="https://github.com/bda-research/node-crawler">
    <img alt="Node.js" src="https://raw.githubusercontent.com/bda-research/node-crawler/master/crawler_primary.png" width="400"/>
  </a>
</p>

#

[![npm package](https://nodei.co/npm/crawler.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/crawler/)

[![build status](https://travis-ci.org/bda-research/node-crawler.svg?branch=master)](https://travis-ci.org/bda-research/node-crawler)
[![Coverage Status](https://coveralls.io/repos/github/bda-research/node-crawler/badge.svg?branch=master)](https://coveralls.io/github/bda-research/node-crawler?branch=master)
[![Dependency Status](https://david-dm.org/bda-research/node-crawler/status.svg)](https://david-dm.org/bda-research/node-crawler)
[![NPM download][download-image]][download-url]
[![NPM quality][quality-image]][quality-url]
[![Gitter](https://img.shields.io/badge/gitter-join_chat-blue.svg?style=flat-square)](https://gitter.im/node-crawler/discuss?utm_source=badge)

[quality-image]: http://npm.packagequality.com/shield/crawler.svg?style=flat-square
[quality-url]: http://packagequality.com/#?package=crawler
[download-image]: https://img.shields.io/npm/dm/crawler.svg?style=flat-square
[download-url]: https://npmjs.org/package/crawler

Most powerful, popular and production crawling/scraping package for Node, happy hacking :)

Features:

- server-side DOM & automatic jQuery insertion with Cheerio (default) or JSDOM
- Configurable pool size and retries
- Control rate limit
- Priority queue of requests
- forceUTF8 mode to let crawler deal for you with charset detection and conversion
- Compatible with 4.x or newer version

Here is the [CHANGELOG](https://github.com/bda-research/node-crawler/blob/master/CHANGELOG.md)

Thanks to [Authuir](https://github.com/authuir), we have a [Chinese](http://node-crawler.readthedocs.io/zh_CN/latest/) docs. Other languages are welcomed!

# Table of Contents

- [Get started](#get-started)
  - [Install](#install)
  - [Basic usage](#basic-usage)
  - [Slow down](#slow-down)
  - [Custom parameters](#custom-parameters)
  - [Raw body](#raw-body)
- [Advanced](#advanced)
  - [Work with bottleneck](#work-with-bottleneck)
  - [Class:Crawler](#classcrawler)
    - [Event: 'schedule'](#event-schedule)
    - [Event: 'request'](#event-request)
    - [Event: 'drain'](#event-drain)
    - [crawler.queue(uri|options)](#crawlerqueueurioptions)
    - [crawler.queueSize](#crawlerqueuesize)
  - [Options reference](#options-reference)
    - [Basic request options](#basic-request-options)
    - [Callbacks](#callbacks)
    - [Schedule options](#schedule-options)
    - [Retry options](#retry-options)
    - [Server-side DOM options](#server-side-dom-options)
    - [Charset encoding](#charset-encoding)
    - [Cache](#cache)
    - [Http headers](#http-headers)
  - [Work with Cheerio or JSDOM](#work-with-cheerio-or-jsdom)
    - [Working with Cheerio](#working-with-cheerio)
    - [Work with JSDOM](#work-with-jsdom)
- [How to test](#how-to-test)
  - [Alternative: Docker](#alternative-docker)
- [Rough todolist](#rough-todolist)

# Get started

## Install

```sh
$ npm install crawler
```

## Basic usage

```js
var Crawler = require('crawler');

var c = new Crawler({
  maxConnections: 10,
  // This will be called for each crawled page
  callback: function (error, res, done) {
    if (error) {
      console.log(error);
    } else {
      var $ = res.$;
      // $ is Cheerio by default
      //a lean implementation of core jQuery designed specifically for the server
      console.log($('title').text());
    }
    done();
  }
});

// Queue just one URL, with default callback
c.queue('http://www.amazon.com');

// Queue a list of URLs
c.queue(['http://www.google.com/', 'http://www.yahoo.com']);

// Queue URLs with custom callbacks & parameters
c.queue([
  {
    uri: 'http://parishackers.org/',
    jQuery: false,

    // The global callback won't be called
    callback: function (error, res, done) {
      if (error) {
        console.log(error);
      } else {
        console.log('Grabbed', res.body.length, 'bytes');
      }
      done();
    }
  }
]);

// Queue some HTML code directly without grabbing (mostly for tests)
c.queue([
  {
    html: '<p>This is a <strong>test</strong></p>'
  }
]);
```

## Slow down

Use `rateLimit` to slow down when you are visiting web sites.

```js
var Crawler = require('crawler');

var c = new Crawler({
  rateLimit: 1000, // `maxConnections` will be forced to 1
  callback: function (err, res, done) {
    console.log(res.$('title').text());
    done();
  }
});

c.queue(tasks); //between two tasks, minimum time gap is 1000 (ms)
```

## Custom parameters

Sometimes you have to access variables from previous request/response session, what should you do is passing parameters as same as options:

```js
c.queue({
  uri: 'http://www.google.com',
  parameter1: 'value1',
  parameter2: 'value2',
  parameter3: 'value3'
});
```

then access them in callback via `res.options`

```js
console.log(res.options.parameter1);
```

Crawler picks options only needed by request, so don't worry about the redundance.

## Raw body

If you are downloading files like image, pdf, word etc, you have to save the raw response body which means Crawler shouldn't convert it to string. To make it happen, you need to set encoding to null

```js
var Crawler = require('crawler');
var fs = require('fs');

var c = new Crawler({
  encoding: null,
  jQuery: false, // set false to suppress warning message.
  callback: function (err, res, done) {
    if (err) {
      console.error(err.stack);
    } else {
      fs.createWriteStream(res.options.filename).write(res.body);
    }

    done();
  }
});

c.queue({
  uri: 'https://nodejs.org/static/images/logos/nodejs-1920x1200.png',
  filename: 'nodejs-1920x1200.png'
});
```

# Advanced

## Work with bottleneck

Control rate limit for with limiter. All tasks submit to a limiter will abide the `rateLimit` and `maxConnections` restrictions of the limiter. `rateLimit` is the minimum time gap between two tasks. `maxConnections` is the maximum number of tasks that can be running at the same time. Limiters are independent of each other. One common use case is setting different limiters for different proxies. One thing is worth noticing, when `rateLimit` is set to a non-zero value, `maxConnections` will be forced to 1.

```js
var crawler = require('crawler');

var c = new Crawler({
  rateLimit: 2000,
  maxConnections: 1,
  callback: function (error, res, done) {
    if (error) {
      console.log(error);
    } else {
      var $ = res.$;
      console.log($('title').text());
    }
    done();
  }
});

// if you want to crawl some website with 2000ms gap between requests
c.queue('http://www.somewebsite.com/page/1');
c.queue('http://www.somewebsite.com/page/2');
c.queue('http://www.somewebsite.com/page/3');

// if you want to crawl some website using proxy with 2000ms gap between requests for each proxy
c.queue({
  uri: 'http://www.somewebsite.com/page/1',
  limiter: 'proxy_1',
  proxy: 'proxy_1'
});
c.queue({
  uri: 'http://www.somewebsite.com/page/2',
  limiter: 'proxy_2',
  proxy: 'proxy_2'
});
c.queue({
  uri: 'http://www.somewebsite.com/page/3',
  limiter: 'proxy_3',
  proxy: 'proxy_3'
});
c.queue({
  uri: 'http://www.somewebsite.com/page/4',
  limiter: 'proxy_1',
  proxy: 'proxy_1'
});
```

## Class:Crawler

### Event: 'schedule'

- `options` [Options](#options-reference)

Emitted when a task is being added to scheduler.

```js
crawler.on('schedule', function (options) {
  options.proxy = 'http://proxy:port';
});
```

Emitted when limiter has been changed.

### Event: 'request'

- `options` [Options](#options-reference)

Emitted when crawler is ready to send a request.

If you are going to modify options at last stage before requesting, just listen on it.

```js
crawler.on('request', function (options) {
  options.qs.timestamp = new Date().getTime();
});
```

### Event: 'drain'

Emitted when queue is empty.

```js
crawler.on('drain', function () {
  // For example, release a connection to database.
  db.end(); // close connection to MySQL
});
```

### crawler.queue(uri|options)

- `uri` [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type)
- `options` [Options](#options-reference)

Enqueue a task and wait for it to be executed.

### crawler.queueSize

- [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type)

Size of queue, read-only

## Options reference

You can pass these options to the Crawler() constructor if you want them to be global or as
items in the queue() calls if you want them to be specific to that item (overwriting global options)

This options list is a strict superset of [mikeal's request options](https://github.com/mikeal/request#requestoptions-callback) and will be directly passed to
the request() method.

### Basic request options

- `options.uri`: [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type) The url you want to crawl.
- `options.timeout` : [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) In milliseconds (Default 15000).
- [All mikeal's request options are accepted](https://github.com/mikeal/request#requestoptions-callback).

### Callbacks

- `callback(error, res, done)`: Function that will be called after a request was completed
  - `error`: [Error](https://nodejs.org/api/errors.html)
  - `res`: [http.IncomingMessage](https://nodejs.org/api/http.html#http_class_http_incomingmessage) A response of standard IncomingMessage includes `$` and `options`
    - `res.statusCode`: [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) HTTP status code. E.G.`200`
    - `res.body`: [Buffer](https://nodejs.org/api/buffer.html) | [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type) HTTP response content which could be a html page, plain text or xml document e.g.
    - `res.headers`: [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) HTTP response headers
    - `res.request`: [Request](https://github.com/request/request) An instance of Mikeal's `Request` instead of [http.ClientRequest](https://nodejs.org/api/http.html#http_class_http_clientrequest)
      - `res.request.uri`: [urlObject](https://nodejs.org/api/url.html#url_url_strings_and_url_objects) HTTP request entity of parsed url
      - `res.request.method`: [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type) HTTP request method. E.G. `GET`
      - `res.request.headers`: [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) HTTP request headers
    - `res.options`: [Options](#options-reference) of this task
    - `$`: [jQuery Selector](https://api.jquery.com/category/selectors/) A selector for html or xml document.
  - `done`: [Function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function) It must be called when you've done your work in callback.

### Schedule options

- `options.maxConnections`: [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) Size of the worker pool (Default 10).
- `options.rateLimit`: [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) Number of milliseconds to delay between each requests (Default 0).
- `options.priority`: [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) Priority of this request (Default 5). Low values have higher priority.

### Retry options

- `options.retries`: [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) Number of retries if the request fails (Default 3),
- `options.retryTimeout`: [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) Number of milliseconds to wait before retrying (Default 10000),

### Server-side DOM options

- `options.jQuery`: [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type)|[String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type)|[Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) Use `cheerio` with default configurations to inject document if true or "cheerio". Or use customized `cheerio` if an object with [Parser options](https://github.com/fb55/htmlparser2/wiki/Parser-options). Disable injecting jQuery selector if false. If you have memory leak issue in your project, use "whacko", an alternative parser,to avoid that. (Default true)

### Charset encoding

- `options.forceUTF8`: [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) If true crawler will get charset from HTTP headers or meta tag in html and convert it to UTF8 if necessary. Never worry about encoding anymore! (Default true),
- `options.incomingEncoding`: [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type) With forceUTF8: true to set encoding manually (Default null) so that crawler will not have to detect charset by itself. For example, `incomingEncoding : 'windows-1255'`. See [all supported encodings](https://github.com/ashtuchkin/iconv-lite/wiki/Supported-Encodings)

### Cache

- `options.skipDuplicates`: [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) If true skips URIs that were already crawled, without even calling callback() (Default false). **This is not recommended**, it's better to handle outside `Crawler` use [seenreq](https://github.com/mike442144/seenreq)

### Http headers

- `options.rotateUA`: [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) If true, `userAgent` should be an array and rotate it (Default false)
- `options.userAgent`: [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type)|[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array), If `rotateUA` is false, but `userAgent` is an array, crawler will use the first one.
- `options.referer`: [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type) If truthy sets the HTTP referer header
- `options.removeRefererHeader`: [Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) If true preserves the set referer during redirects
- `options.headers`: [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) Raw key-value of http headers

### Https socks5

```js
const Agent = require('socks5-https-client/lib/Agent');
//...
var c = new Crawler({
  // rateLimit: 2000,
  maxConnections: 20,
  agentClass: Agent, //adding socks5 https agent
  method: 'GET',
  strictSSL: true,
  agentOptions: {
    socksHost: 'localhost',
    socksPort: 9050
  },
  // debug: true,
  callback: function (error, res, done) {
    if (error) {
      console.log(error);
    } else {
      //
    }
    done();
  }
});
```

## Work with Cheerio or JSDOM

Crawler by default use [Cheerio](https://github.com/cheeriojs/cheerio) instead of [JSDOM](https://github.com/tmpvar/jsdom). JSDOM is more robust, if you want to use JSDOM you will have to require it `require('jsdom')` in your own script before passing it to crawler.

### Working with Cheerio

```js
jQuery: true //(default)
//OR
jQuery: 'cheerio'
//OR
jQuery: {
    name: 'cheerio',
    options: {
        normalizeWhitespace: true,
        xmlMode: true
    }
}
```

These parsing options are taken directly from [htmlparser2](https://github.com/fb55/htmlparser2/wiki/Parser-options), therefore any options that can be used in `htmlparser2` are valid in cheerio as well. The default options are:

```js
{
    normalizeWhitespace: false,
    xmlMode: false,
    decodeEntities: true
}
```

For a full list of options and their effects, see [this](https://github.com/fb55/DomHandler) and

CHANGELOG.md
## node-webcrawler ChangeLog

1.2.1

- [#310](https://github.com/bda-research/node-crawler/issues/310) Upgrade dependencies' version(@mike442144)
- [#303](https://github.com/bda-research/node-crawler/issues/303) Update seenreq to v3(@mike442144)
- [#304](https://github.com/bda-research/node-crawler/pull/304) Replacement of istanbul with nyc (@kossidts)
- [#300](https://github.com/bda-research/node-crawler/pull/300) Add formData arg to requestArgs (@humandevmode)
- [#280](https://github.com/bda-research/node-crawler/pull/280) 20180611 updatetestwithnock (@Dong-Gao)

  1.2.0

- [#278](https://github.com/bda-research/node-crawler/pull/278) Added filestream require to download section (@swosko)
- Use `nock` to mock testing instead of httpbin
- Replace jshint by eslint
- Fix code to pass eslint rules

  1.1.4

- Tolerate incorrect `Content-Type` header [#270](https://github.com/bda-research/node-crawler/pull/270), [#193](https://github.com/bda-research/node-crawler/issues/193)
- Added examples [#272](https://github.com/bda-research/node-crawler/pull/272), [267](https://github.com/bda-research/node-crawler/issues/267)
- Fixed "skipDuplicates" and "retries" config incompatible bug [#261](https://github.com/bda-research/node-crawler/issues/261)
- Fix typo in README [#268](https://github.com/bda-research/node-crawler/pull/268)

  1.1.3

- Upgraded `request.js` and `lodash`

  1.1.2

- Recognize all XML MIME types to inject jQuery [#245](https://github.com/bda-research/node-crawler/pull/245)
- Allow options to specify the Agent for Request [#246](https://github.com/bda-research/node-crawler/pull/246)
- Added logo

  1.1.1

- added a way to replace the global options.headers keys by queuing options.headers [#241](https://github.com/bda-research/node-crawler/issues/241)
- fix bug of using last jar object if current options doesn't contain `jar` option [#240](https://github.com/bda-research/node-crawler/issues/240)
- fix bug of encoding [#233](https://github.com/bda-research/node-crawler/issues/233)
- added seenreq options [#208](https://github.com/bda-research/node-crawler/issues/208)
- added preRequest, setLimiterProperty, direct request functions

  1.0.5

- fix missing debugging messages [#213](https://github.com/bda-research/node-crawler/issues/213)
- fix bug of 'drain' never called [#210](https://github.com/bda-research/node-crawler/issues/210)

  1.0.4

- fix bug of charset detecting [#203](https://github.com/bda-research/node-crawler/issues/203)
- keep node version up to date in travis scripts

  1.0.3

- fix bug, skipDuplicate and rotateUA don't work even if set true

  1.0.0

- upgrade jsdom up to 9.6.x
- remove 0.10 and 0.12 support [#170](https://github.com/bda-research/node-crawler/issues/170)
- control dependencies version using ^ and ~ [#169](https://github.com/bda-research/node-crawler/issues/169)
- remove node-pool
- notify bottleneck until a task is completed
- replace bottleneck by bottleneckp, which has priority
- change default log function
- use event listener on `request` and `drain` instead of global function [#144](https://github.com/bda-research/node-crawler/issues/144)
- default set forceUTF8 to true
- detect `ESOCKETTIMEDOUT` instead of `ETIMEDOUT` when timeout in test
- add `done` function in callback to avoid async trap
- do not convert response body to string if `encoding` is null [#118](https://github.com/bda-research/node-crawler/issues/118)
- add result document [#68](https://github.com/bda-research/node-crawler/issues/68) [#116](https://github.com/bda-research/node-crawler/issues/116)
- add event `schedule` which is emitted when a task is being added to scheduler
- in callback, move \$ into `res` because of weird API
- change rateLimits to rateLimit

  0.7.5

- delete entity in options before copy, and assgin after, `jar` is one of the typical properties which is an `Entity` wich functions [#177](https://github.com/bda-research/node-crawler/issues/177)
- upgrade `request` to version 2.74.0

  0.7.4

- change `debug` option to instance level instead of `options`
- update README.md to detail error handling
- call `onDrain` with scope of `this`
- upgrade `seenreq` version to 0.1.7

  0.7.0

- cancel recursion in queue
- upgrade `request` version to v2.67.0

  0.6.9

- use `bottleneckConcurrent` instead of `maxConnections`, default `10000`
- add debug info

  0.6.5

- fix a deep and big bug when initializing Pool, that may lead to sequence execution. [#2](https://github.com/bda-research/node-webcrawler/issues/2)
- print log of Pool status

  0.6.3

- you could also get `result.options` from callback even when some errors ouccurred [#127](https://github.com/bda-research/node-crawler/issues/127) [#86](https://github.com/bda-research/node-crawler/issues/86)
- add test for `bottleneck`

  0.6.0

- add `bottleneck` to implement rate limit, one can set limit for each connection at same time.

  0.5.2

- you can manually terminate all the resources in your pool, when `onDrain` called, before their timeouts have been reached
- add a read-only property `queueSize` to crawler [#148](https://github.com/bda-research/node-crawler/issues/148) [#76](https://github.com/bda-research/node-crawler/issues/76) [#107](https://github.com/bda-research/node-crawler/issues/107)

  0.5.1

- remove cache feature, it's useless
- add `localAddress`, `time`, `tunnel`, `proxyHeaderWhiteList`, `proxyHeaderExclusiveList` properties to pass to `request` [#155](https://github.com/bda-research/node-crawler/issues/155)

  0.5.0

- parse charset from `content-type` in http headers or meta tag in html, then convert
- big5 charset is avaliable as the `iconv-lite` has already supported it
- default enable gzip in request header
- remove unzip code in crawler since `request` will do this
- body will return as a Buffer if encoding is null which is an option in `request`
- remove cache and skip duplicate `request` for `GET`, `POST`(only for type `urlencode`), `HEAD`

lib/crawler.js
'use strict';

const path = require('path'),
  util = require('util'),
  EventEmitter = require('events').EventEmitter,
  request = require('request-promise-native'),
  _ = require('lodash'),
  cheerio = require('cheerio'),
  fs = require('fs'),
  Bottleneck = require('bottleneck'),
  seenreq = require('seenreq'),
  iconvLite = require('iconv-lite'),
  typeis = require('type-is').is;

let whacko = null,
  level;

class CrawlError extends Error {
  constructor(originalError, options) {
    super();
    this.originalError = originalError;
    this.options = options;
  }
}
const levels = [
  'silly',
  'debug',
  'verbose',
  'info',
  'warn',
  'error',
  'critical'
];
try {
  whacko = require('whacko');
} catch (e) {
  e.code;
}

function defaultLog() {
  //2016-11-24T12:22:55.639Z - debug:
  if (levels.indexOf(arguments[0]) >= levels.indexOf(level))
    console.log(
      new Date().toJSON() + ' - ' + arguments[0] + ': CRAWLER %s',
      util.format.apply(util, Array.prototype.slice.call(arguments, 1))
    );
}

function checkJQueryNaming(options) {
  if ('jquery' in options) {
    options.jQuery = options.jquery;
    delete options.jquery;
  }
  return options;
}

function readJqueryUrl(url, callback) {
  if (url.match(/^(file:\/\/|\w+:|\/)/)) {
    fs.readFile(url.replace(/^file:\/\//, ''), 'utf-8', function (err, jq) {
      callback(err, jq);
    });
  } else {
    callback(null, url);
  }
}

function contentType(res) {
  return get(res, 'content-type')
    .split(';')
    .filter((item) => item.trim().length !== 0)
    .join(';');
}

function get(res, field) {
  return res.headers[field.toLowerCase()] || '';
}

let log = defaultLog;

class Crawler extends EventEmitter {
  constructor(options) {
    super();
    options = options || {};
    if (['onDrain', 'cache'].some((key) => key in options)) {
      throw new Error(
        'Support for "onDrain", "cache" has been removed! For more details, see https://github.com/bda-research/node-crawler'
      );
    }

    this.init(options);
  }

  init(options) {
    const defaultOptions = {
      autoWindowClose: true,
      forceUTF8: true,
      gzip: true,
      incomingEncoding: null,
      jQuery: true,
      maxConnections: 10,
      method: 'GET',
      priority: 5,
      rateLimit: 0,
      referer: false,
      retries: 3,
      retryTimeout: 10000,
      timeout: 15000,
      skipDuplicates: false,
      rotateUA: false
    };

    //return defaultOptions with overriden properties from options.
    this.options = { ...defaultOptions, ...options };

    // you can use jquery or jQuery
    this.options = checkJQueryNaming(this.options);

    // Don't make these options persist to individual queries
    this.globalOnlyOptions = [
      'maxConnections',
      'rateLimit',
      'skipDuplicates',
      'rotateUA'
    ];

    this.limiters = new Bottleneck({
      maxConcurrent: this.options.maxConnections,
      minTime: this.options.rateLimit
    });

    level = this.options.debug ? 'debug' : 'info';

    if (this.options.logger)
      log = this.options.logger.log.bind(this.options.logger);

    this.log = log;

    this.seen = new seenreq(this.options.seenreq);
    this.seen
      .initialize()
      .then(() => log('debug', 'seenreq is initialized.'))
      .catch((e) => log('error', e));
    this.limiters.on('failed', async (error, jobInfo) => {
      if (
        error instanceof CrawlError &&
        error.options.retries > jobInfo.retryCount
      ) {
        const options = error.options;
        log(
          'error',
          'Error ' +
            error.originalError +
            ' when fetching ' +
            (options.uri || options.url) +
            (options.retries
              ? ' (' + options.retries - jobInfo.retryCount + ' retries left)'
              : '')
        );

        return error.options.retryTimeout;
      }
    });

    this.limiters.on('idle', async () => {
      log('debug', 'Queue size: %d', this.limiters.queued());
      return this.emit('drain');
    });
  }

  _inject(response, options, callback) {
    let $;

    if (options.jQuery === 'whacko') {
      if (!whacko) {
        throw new Error(
          'Please install whacko by your own since `crawler` detected you specify explicitly'
        );
      }

      $ = whacko.load(response.body);
      callback(null, response, options, $);
    } else if (
      options.jQuery === 'cheerio' ||
      options.jQuery.name === 'cheerio' ||
      options.jQuery === true
    ) {
      const defaultCheerioOptions = {
        normalizeWhitespace: false,
        xmlMode: false,
        decodeEntities: true
      };
      const cheerioOptions = options.jQuery.options || defaultCheerioOptions;
      $ = cheerio.load(response.body, cheerioOptions);

      callback(null, response, options, $);
    } else if (options.jQuery.jsdom) {
      const jsdom = options.jQuery.jsdom;
      const scriptLocation = path.resolve(
        __dirname,
        '../vendor/jquery-2.1.1.min.js'
      );

      //Use promises
      readJqueryUrl(scriptLocation, (err, jquery) => {
        try {
          jsdom.env({
            url: options.uri,
            html: response.body,
            src: [jquery],
            done: (errors, window) => {
              $ = window.jQuery;
              callback(errors, response, options, $);

              try {
                window.close();
                window = null;
              } catch (err) {
                log('error', err);
              }
            }
          });
        } catch (e) {
          options.callback(e, { options }, options.release);
        }
      });
    }
    // Jquery is set to false are not set
    else {
      callback(null, response, options);
    }
  }

  isIllegal(options) {
    return (
      _.isNull(options) ||
      _.isUndefined(options) ||
      (!_.isString(options) && !_.isPlainObject(options))
    );
  }

  queue(options) {
    // Did you get a single object or string? Make it compatible.
    options = _.isArray(options) ? options : [options];

    options = _.flattenDeep(options);

    for (let i = 0; i < options.length; ++i) {
      if (this.isIllegal(options[i])) {
        log('warn', 'Illegal queue option: ', JSON.stringify(options[i]));
        continue;
      }
      this._pushToQueue(
        _.isString(options[i]) ? { uri: options[i] } : options[i]
      );
    }
  }

  get queueSize() {
    return this.limiters.queued();
  }

  _pushToQueue(options) {
    // you can use jquery or jQuery
    options = checkJQueryNaming(options);

    _.defaults(options, this.options);
    options.headers = { ...this.options.headers, ...options.headers };

    // Remove all the global options from our options
    // TODO we are doing this for every _pushToQueue, find a way to avoid this
    this.globalOnlyOptions.forEach(
      (globalOnlyOption) => delete options[globalOnlyOption]
    );

    // If duplicate skipping is enabled, avoid queueing entirely for URLs we already crawled
    if (!this.options.skipDuplicates) {
      this._schedule(options);
      return;
    }

    this.seen
      .exists(options, options.seenreq)
      .then((rst) => {
        if (!rst) {
          this._schedule(options);
        }
      })
      .catch((e) => log('error', e));
  }

  async _schedule(options) {
    this.emit('schedule', options);
    try {
      await this.limiters.schedule(options, async () => {
        const promise = new Promise((resolve) => {
          options.release = () => {
            resolve();
          };
        });

        if (!options.callback) {
          options.callback = options.release;
        }

        if (options.html) {
          this._onContent(null, options, {
            body: options.html,
            headers: { 'content-type': 'text/html' }
          });
        } else if (typeof options.uri === 'function') {
          options.uri(async (uri) => {
            options.uri = uri;
            await this._buildHTTPRequest(options);
          });
        } else {
          await this._buildHTTPRequest(options);
        }
        return promise;
      });
    } catch (error) {
      this._onContent(error.originalError, error.options);
    }
  }

  async _buildHTTPRequest(options) {
    log('debug', options.method + ' ' + options.uri);
    if (options.proxy) log('debug', 'Use proxy: %s', options.proxy);

    // Cloning keeps the opts parameter clean:
    // - some versions of "request" apply the second parameter as a
    // property called "callback" to the first parameter
    // - keeps the query object fresh in case of a retry

    const ropts = { ...options };

    if (!ropts.headers) {
      ropts.headers = {};
    }
    if (ropts.forceUTF8) {
      ropts.encoding = null;
    }
    // specifying json in request will have request sets body to JSON representation of value and
    // adds Content-type: application/json header. Additionally, parses the response body as JSON
    // so the response will be JSON object, no need to deal with encoding
    if (ropts.json) {
      options.encoding = null;
    }
    if (ropts.userAgent) {
      if (this.options.rotateUA && _.isArray(ropts.userAgent)) {
        ropts.headers['User-Agent'] = ropts.userAgent[0];
        // If "rotateUA" is true, rotate User-Agent
        options.userAgent.push(options.userAgent.shift());
      } else {
        ropts.headers['User-Agent'] = ropts.userAgent;
      }
    }

    if (ropts.referer) {
      ropts.headers.Referer = ropts.referer;
    }

    if (ropts.proxies && ropts.proxies.length) {
      ropts.proxy = ropts.proxies[0];
    }

    const doRequest = async () => {
      if (ropts.skipEventRequest !== true) {
        this.emit('request', ropts);
      }

      const requestArgs = [
        'uri',
        'url',
        'qs',
        'method',
        'headers',
        'body',
        'form',
        'formData',
        'json',
        'multipart',
        'followRedirect',
        'followAllRedirects',
        'maxRedirects',
        'removeRefererHeader',
        'encoding',
        'pool',
        'timeout',
        'proxy',
        'auth',
        'oauth',
        'strictSSL',
        'jar',
        'aws',
        'gzip',
        'time',
        'tunnel',
        'proxyHeaderWhiteList',
        'proxyHeaderExclusiveList',
        'localAddress',
        'forever',
        'agent',
        'strictSSL',
        'agentOptions',
        'agentClass'
      ];
      let response;
      try {
        response = await request({
          ..._.pick.apply(this, [ropts].concat(requestArgs)),
          resolveWithFullResponse: true,
          simple: false
        });
      } catch (e) {
        throw new CrawlError(e, options);
      }

      this._onContent(null, options, response);
    };

    await doRequest();
  }

  _onContent(error, options, response) {
    if (error) {
      options.callback(error, { options: options }, options.release);

      return;
    }

    if (!response.body) {
      response.body = '';
    }

    log(
      'debug',
      'Got ' +
        (options.uri || 'html') +
        ' (' +
        response.body.length +
        ' bytes)...'
    );

    try {
      this._doEncoding(options, response);
    } catch (e) {
      log('error', e);
      return options.callback(e, { options: options }, options.release);
    }

    response.options = options;

    if (options.method === 'HEAD' || !options.jQuery) {
      return options.callback(null, response, options.release);
    }

    const injectableTypes = [
      'html',
      'xhtml',
      'text/xml',
      'application/xml',
      '+xml'
    ];
    if (!options.html && !typeis(contentType(response), injectableTypes)) {
      log(
        'warn',
        'response body is not HTML, skip injecting. Set jQuery to false to suppress this message'
      );
      return options.callback(null, response, options.release);
    }

    log('debug', 'Injecting');

    this._inject(response, options, this._injected.bind(this));
  }

  _injected(errors, response, options, $) {
    log('debug', 'Injected');

    response.$ = $;
    options.callback(errors, response, options.release);
  }

  _doEncoding(options, response) {
    if (options.encoding === null) {
      return;
    }

    if (options.forceUTF8) {
      const charset = options.incomingEncoding || this._parseCharset(response);
      response.charset = charset;
      log('debug', 'Charset ' + charset);

      if (charset !== 'utf-8' && charset !== 'ascii') {
        // convert response.body into 'utf-8' encoded buffer
        response.body = iconvLite.decode(response.body, charset);
      }
    }

    response.body = response.body.toString();
  }

  _parseCharset(res) {
    //Browsers treat gb2312 as gbk, but iconv-lite not.
    //Replace gb2312 with gbk, in order to parse the pages which say gb2312 but actually are gbk.
    const getCharset = (str) => {
      const charset = ((str && str.match(/charset=['"]?([\w.-]+)/i)) || [
        0,
        null
      ])[1];
      return charset && charset.replace(/:\d{4}$|[^0-9a-z]/g, '') == 'gb2312'
        ? 'gbk'
        : charset;
    };
    const charsetParser = (header, binary, default_charset = null) => {
      return getCharset(header) || getCharset(binary) || default_charset;
    };

    var charset = charsetParser(contentType(res));
    if (charset) return charset;

    if (!typeis(contentType(res), ['html'])) {
      log(
        'debug',
        'Charset not detected in response headers, please specify using `incomingEncoding`, use `utf-8` by default'
      );
      return 'utf-8';
    }

    const body = res.body instanceof Buffer ? res.body.toString() : res.body;
    charset = charsetParser(contentType(res), body, 'utf-8');

    return charset;
  }
}

module.exports = Crawler;
